### ISSUE TITLE
Multi-controls/Multi-pipes

### ISSUE THREAD

**Issue by 2opremio** [weaveworks/scope#1050]
For terminals: allow creating synched terminals à la tmux for multiple containers/hosts.

<img width="2368" alt="screen shot 2016-02-26 at 16 12 10" src="https://cloud.githubusercontent.com/assets/2362916/13357748/bed1b4f8-dca3-11e5-9786-1fd03121034c.png">

For logs: show the interspersed logs of multiple containers/hosts in the same window

This will greatly simplify doing diagnostics/administering tasks:

* It would allow me to completely ditch ssh
* It would be particularly useful when load balancing to multiple containers, in which seeing the logs/terminal of a single container doesn't give you the full picture.

**Comment by 2opremio** [comment]
Related #1049

**Comment by peterbourgon** [comment]
Terminals: strong :-1:, Scope is not and should not be an ersatz terminal emulator, the security implications alone are horrifying

Logs: :+1:, "attach to multiple and mux" seems like a good and coherent extension of current attach functionality

**Comment by 2opremio** [comment]
> should not

Let's try to keep an open mind please and not talk in absolute terms, please.

**Comment by 2opremio** [comment]
> the security implications alone are horrifying

Not considerably worse than what we have today, we allow terminals in priviledged containers plus, even in the case of the unpriviledged ones, the security guarantees offered by containers are not very strong.

Also, I am just trying to be realistic. There's administrative tasks for which scope is not going to be enough and you will need a terminal. In my perspective, making this simple and easy adds a lot of value.

**Comment by peterbourgon** [comment]
> Not considerably worse than what we have today

What we have today is horrifying in equal measure. I hope we all recognize that and will work to limit and scale back, rather than enhance, that horror.

**Comment by 2opremio** [comment]
> What we have today is horrifying in equal measure. I hope we all recognize that and will work to limit and scale back, rather than enhance, that horror.

I beg to differ, I've found it very useful when administering the service.

**Comment by peterbourgon** [comment]
>  I've found it very useful when administering the service.

Of course it is useful. To you, and to anyone who happens to gain access to any host on port 4040 :)

**Comment by 2opremio** [comment]
> Of course it is useful. To you, and to anyone who happens to gain access to any host on port 4040 :)

Agreed. We should work on providing better security.

**Comment by monadic** [comment]
Please can we discuss this on Monday

**Comment by 2opremio** [comment]
Absolutely

**Comment by peterbourgon** [comment]
For the record we haven't resolved the points of contention here. I still feel strongly that multiple terminals as described in the OP would be an antifeature. Please don't begin implementation before we can have a roundtable on it :)

**Comment by 2opremio** [comment]
I am going to begin with the host logs, which are not controversial.

**Comment by peterbourgon** [comment]
:+1:

**Comment by rade** [comment]
You can already have multiple terminals in scope, by opening them in turn and using the "detach" button.

So what are we actually missing?

**Comment by 2opremio** [comment]
> So what are we actually missing?

Those terminals are not synched nor tiled and we don't (yet) have host terminals.

The same applies to container/host logs.

**Comment by rade** [comment]
> Those terminals are not synched

What does "synced" mean?

> nor tiled

IMO we should avoid creating a window manager in scope. If you want tiled windows then use a tiling window manager to do that for your detached scope terminal windows.

> we don't (yet) have host terminals.

That is #1049. Not this issue.

> In the container/host logs case you also want them interspersed.

That should be a separate issue. And presumably "interspersed" means there would only be *one* terminal.

**Comment by 2opremio** [comment]
> What does "synced" mean?

The same keyboard input is sent to all terminals, which is very useful to diagnose problems across a group of machines (containers).

> IMO we should avoid creating a window manager in scope. If you want tiled windows then use a tiling window manager to do that for your detached scope terminal windows.

As a user, that extra requirement/complexity of managing the detached windows (e.g. by installing a tiling window manager) wouldn't justify switching from my usual tmux setup (but this may be just me ...).

To clarify this further, my idea is to have a single terminal button for groups of containers/hosts (e.g. containers by image, all hosts ...) which would spawn all terminals at once. In order to provide this, Scope doesn't need to support custom tiling.

> And presumably "interspersed" means there would only be one terminal.

Yes
